### PR TITLE
Feature - FactoryBoy to generate Mock data from Avro Protocols

### PR DESCRIPTION
**Pull request by Antonior26**
I create an example with Variant, this will be how you use it:
* GA4GHVariantFactory()
* GA4GHVariantFactory(referenceBases='T')
* GA4GHVariantFactory.create_batch(10)
* GA4GHVariantFactory.create_batch(10, referenceBases='T')

You can find a extend documentation in https://factoryboy.readthedocs.io/en/latest/index.html

### CODE REVIEW

File: protocols/util/avro_factory.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        if hasattr(attrs['Meta'], 'model'):
+            fuzzy_attributes = mock_schema(attrs['Meta'].model)
+            for parent in args[1]:
+                fuzzy_attributes.update(parent._meta.pre_declarations.declarations)
+            fuzzy_attributes.update(attrs)
+            attrs = fuzzy_attributes
+
+        return super(FactoryAvroMetaClass, mcs).__new__(mcs, args[0], args[1], attrs)
+
+FactoryAvro = FactoryAvroMetaClass(str('Factory'), (BaseFactory,), {
```
Comment (greglever) [inline protocols/util/avro_factory.py]: why are we converting a string to a string here ?

File: protocols/util/avro_factory.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def __new__(mcs, *args):
+        attrs = args[2]
+        if hasattr(attrs['Meta'], 'model'):
+            fuzzy_attributes = mock_schema(attrs['Meta'].model)
+            for parent in args[1]:
+                fuzzy_attributes.update(parent._meta.pre_declarations.declarations)
+            fuzzy_attributes.update(attrs)
+            attrs = fuzzy_attributes
+
+        return super(FactoryAvroMetaClass, mcs).__new__(mcs, args[0], args[1], attrs)
```
Comment (greglever) [inline protocols/util/avro_factory.py]: will `args` always be at least of length 3 ?

File: protocols/util/avro_factory.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def mock_schema(protocol):
+    """
+    :type protocol: Protocol
+    """
+    INT_MIN_VALUE = -(1 << 31)
+    INT_MAX_VALUE = (1 << 31) - 1
+    LONG_MIN_VALUE = -(1 << 63)
+    LONG_MAX_VALUE = (1 << 63) - 1
```
Comment (greglever) [inline protocols/util/avro_factory.py]: I would use:
```
INT_MIN_VALUE = -sys.maxint-1
INT_MAX_VALUE = sys.maxint
LONG_MIN_VALUE = -sys.maxsize-1
LONG_MAX_VALUE = sys.maxsize
```
just for readability

File: protocols/util/avro_factory.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import sys
+import factory.fuzzy
+from factory.base import FactoryMetaClass, BaseFactory, BaseMeta
```
Comment (greglever) [inline protocols/util/avro_factory.py]: Can we have the builtin imports then the third party imports, so:
```
import sys

import factory.fuzzy
from factory.base import BaseMeta
from factory.base import BaseFactory
from factory.base import FactoryMetaClass
```

File: protocols/util/factories.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import factory.fuzzy
+import itertools
+from protocols.ga4gh_3_0_0 import Variant, Call
+from .avro_factory import FactoryAvro
```
Comment (greglever) [inline protocols/util/factories.py]: Can we please have builtins then third party then internal imports ? So something like:
```
import itertools

import factory.fuzzy

from .avro_factory import FactoryAvro
from protocols.ga4gh_3_0_0 import Call
from protocols.ga4gh_3_0_0 import Variant
```

File: protocols/util/avro_factory.py (hunk 4 above)
Comment (Antonior26) [inline protocols/util/avro_factory.py]: :+1:

File: protocols/util/avro_factory.py (hunk 2 above)
Comment (Antonior26) [inline protocols/util/avro_factory.py]: yes the method is called in line 50 (fixed, always 3 args)

File: protocols/util/avro_factory.py (hunk 1 above)
Comment (Antonior26) [inline protocols/util/avro_factory.py]: to avoid issues with Py2's type() expecting bytes instead of unicode.

Comment (Antonior26) [inline protocols/util/avro_factory.py]: (From the original code), I just copied their strategy to create the FactoryMetaClass
https://github.com/FactoryBoy/factory_boy/blob/master/factory/base.py#L670

File: protocols/util/avro_factory.py (hunk 3 above)
Comment (Antonior26) [inline protocols/util/avro_factory.py]: I copied from your code :P

File: protocols/util/avro_factory.py (hunk 2 above)
Comment (greglever) [inline protocols/util/avro_factory.py]: great, thanks for clearing that up @Antonior26

File: protocols/util/avro_factory.py (hunk 1 above)
Comment (greglever) [inline protocols/util/avro_factory.py]: ah great, thanks for clearing that up @Antonior26

File: protocols/util/avro_factory.py (hunk 3 above)
Comment (greglever) [inline protocols/util/avro_factory.py]: @Antonior26 - ah fair if it's just a copy and paste then keep it the same, at least we know it works that way

Comment (priesgo) [inline protocols/util/avro_factory.py]: I remember commenting this to you @greglever and you defended the above!!!

Comment (greglever) [inline protocols/util/avro_factory.py]: haha, @priesgo I didn't realise it was copied from exactly the same code. We should really refactor all of that soon anyway.

File: protocols/util/avro_factory.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        elif schema_type == 'bytes':
+            mock_result[name] = b'a'
+        elif schema_type == 'int':
+            mock_result[name] = factory.fuzzy.FuzzyInteger(INT_MIN_VALUE, INT_MAX_VALUE)
+        elif schema_type == 'long':
+            mock_result[name] = factory.fuzzy.FuzzyInteger(LONG_MIN_VALUE, LONG_MAX_VALUE)
+        elif schema_type in ['float', 'double']:
+            mock_result[name] = factory.fuzzy.FuzzyFloat(sys.float_info.min)
+        elif schema_type == 'enum':
+            mock_result[name] = factory.fuzzy.FuzzyChoice(expected_schema.symbols)
```
Comment (priesgo) [inline protocols/util/avro_factory.py]: When you add support for arrays, the podam library we used in Java creates by default all lists of length 5. If you want to customise length of a given array you can only specify in an ad hoc parameter in kwargs.

File: protocols/util/factories.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class GA4GHVariantFactory(FactoryAvro):
+    def __init__(self, *args, **kwargs):
+        super(GA4GHVariantFactory, self).__init__(*args, **kwargs)
+
+    class Meta:
+        model = Variant
+    start = factory.fuzzy.FuzzyInteger(1, 10000000)
```
Comment (priesgo) [inline protocols/util/factories.py]: Can we use kwargs to pass ad hoc parameters???
I was thinking GA4GHVariantFactory(start=123456789, referenceBases='T', alternateBases='G', referenceName='22')

Comment (Antonior26) [inline protocols/util/factories.py]: Yes indeed!

Comment (Antonior26) [inline protocols/util/factories.py]: I will prepare some test, trying to do complex things, in the meantime https://factoryboy.readthedocs.io/en/latest/index.html
 and @gerlv has experience with the library (you can check panelApp tests)

File: protocols/tests/test_util/test_factories.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+from unittest import TestCase
+
+import factory
+
+from models.ga4ghvariant import GA4GHVariant
+from protocols.util.factories import GA4GHVariantFactory, CallFactory
```
Comment (greglever) [inline protocols/tests/test_util/test_factories.py]: :+1:

**Pull request merged**